### PR TITLE
gdrive: add open

### LINKED ISSUE #3408: api: support streaming from Google Drive remotes
> Extracted from https://github.com/iterative/dvc.org/pull/908#discussion_r382941301

GDrive is the only remote type that doesn't support file streaming. Should we work on adding it?

- [ ] and update docs (e.g. api ref)

### LINKED ISSUE #3897: "AttributeError: NoneType object has no attribute 'wdir'" on dvc get-url call
## Bug Report

When calling ``dvc get-url gdrive://{mylink}`` from the command line, I am getting the following AttributeError: "NoneType' object has no attribute 'wdir'.  The full traceback is below.

I'm calling ``dvc get-url`` from outside the .dvc working directory.  A quick look at the source code suggests that this is related to output.LocalOutput being initialized with stage=None from the dependency.loads_from() function in repo.get_url.py.

### Please provide information about your setup
**Output of `dvc version`:**
```sh
$ dvc version
DVC version: 0.94.0
Python version: 3.7.6
Platform: Windows-10-10.0.18362-SP0
Binary: False
Package: pip
Supported remotes: gdrive, http, https
Filesystem type (workspace): ('NTFS', 'C:\\')
```

**Additional Information (if any):**

```
Traceback (most recent call last):
  File "d:\miniconda3\lib\site-packages\dvc\main.py", line 49, in main
    ret = cmd.run()
  File "d:\miniconda3\lib\site-packages\dvc\command\get_url.py", line 17, in run
    Repo.get_url(self.args.url, out=self.args.out)
  File "d:\miniconda3\lib\site-packages\dvc\repo\get_url.py", line 17, in get_url
    (dep,) = dependency.loads_from(None, [url])
  File "d:\miniconda3\lib\site-packages\dvc\dependency\__init__.py", line 83, in loads_from
    ret.append(_get(stage, s, info))
  File "d:\miniconda3\lib\site-packages\dvc\dependency\__init__.py", line 68, in _get
    return LocalDependency(stage, p, info)
  File "d:\miniconda3\lib\site-packages\dvc\output\local.py", line 25, in __init__
    super().__init__(stage, path, *args, **kwargs)
  File "d:\miniconda3\lib\site-packages\dvc\output\base.py", line 98, in __init__
    self.path_info = self._parse_path(remote, path)
  File "d:\miniconda3\lib\site-packages\dvc\output\local.py", line 40, in _parse_path
    p = self.stage.wdir / p
AttributeError: 'NoneType' object has no attribute 'wdir'
```

### PR DESCRIPTION
**Pull request by casperdcl**
- [x] Add `open()` for `gdrive` remote
- [ ] update docs
- fixes #3408
- related #2865
- related #3897
- [x] depends https://github.com/iterative/PyDrive2/pull/39
  - [x] depends https://github.com/conda-forge/pydrive2-feedstock/pull/13
- [x] ~~depends https://github.com/iterative/PyDrive2/pull/40~~
- [x] tests in https://github.com/iterative/dvc/blob/1.0.0a6/tests/func/test_api.py#L60
  + [x] fix credentials https://travis-ci.com/github/iterative/dvc/jobs/341956240
  + [x] rebase onto current `master`

### CODE REVIEW

File: dvc/dependency/__init__.py [hunk 1]
```diff
@@ -19,6 +20,7 @@
 
 DEPS = [
     AzureDependency,
+    GDriveDependency,
```
Comment (shcheklein) [inline dvc/dependency/__init__.py:23]: is it actually needed to support open?

File: dvc/remote/gdrive.py [hunk 2]
```diff
@@ -393,6 +396,23 @@ def _gdrive_download_file(
         ) as pbar:
             gdrive_file.GetContentFile(to_file, callback=pbar.update_to)
 
+    @contextmanager
```
Comment (shcheklein) [inline dvc/remote/gdrive.py:399]: just to double check - do we properly pass `close` down to the PyDrive?

File: dvc/remote/gdrive.py [hunk 3]
```diff
@@ -393,6 +396,23 @@ def _gdrive_download_file(
...
+    @_gdrive_retry
+    def open(self, path_info, mode="r", encoding=None):
+        assert mode in {"r", "rt", "rb"}
```
Comment (shcheklein) [inline dvc/remote/gdrive.py:402]: should we raise an illegalargument here?

File: dvc/dependency/__init__.py (hunk 1 above)
Comment (casperdcl) [reply dvc/dependency/__init__.py:23]: No - may have to revert `dependency` things for a different PR depending on @efiop feedback.

File: dvc/remote/gdrive.py (hunk 3 above)
Comment (casperdcl) [reply dvc/remote/gdrive.py:402]: It's not done in other places in the code - would have to do everywhere if needed

File: dvc/remote/gdrive.py (hunk 2 above)
Comment (casperdcl) [reply dvc/remote/gdrive.py:399]: pydrive2 doesn't expose any `close()` methods anywhere. I think the underlying `httplib2` closes connections before raising errors; so no special handling needed.

File: dvc/dependency/__init__.py (hunk 1 above)
Comment (efiop) [reply dvc/dependency/__init__.py:23]: @casperdcl `get_file_checksum` is not implemented for GdriveRemote, so this is not needed.

File: tests/func/test_api.py [hunk 4]
```diff
@@ -56,7 +56,9 @@ def test_get_url_requires_dvc(tmp_dir, scm):
         api.get_url("foo", repo=f"file://{tmp_dir}")
 
 
-@pytest.mark.parametrize("remote_url", all_remote_params, indirect=True)
+@pytest.mark.parametrize(
+    "remote_url", all_remote_params + [GDrive], indirect=True
```
Comment (efiop) [inline tests/func/test_api.py:60]: Unless I'm missing something, this change should make all the tests in this test_api.py work with Gdrive.

Comment (casperdcl) [reply tests/func/test_api.py:60]: nope because of https://github.com/iterative/dvc/blob/edfa1741db506ba12ee0244452349c1db059a026/tests/remotes.py#L144 requiring an argument (`self`).

File: tests/func/test_api.py [hunk 5]
```diff
@@ -8,7 +8,7 @@
 from dvc.main import main
 from dvc.path_info import URLInfo
 from dvc.utils.fs import remove
-from tests.remotes import GCP, HDFS, OSS, S3, SSH, Azure, Local
+from tests.remotes import GCP, HDFS, OSS, S3, SSH, Azure, GDrive, Local
 
 remote_params = [S3, GCP, Azure, OSS, SSH, HDFS]
```
Comment (efiop) [inline tests/func/test_api.py:13]: We should be able to simply add Gdrive to this list. Sorry for not mentioning it explicitly earlier.

File: tests/func/test_api.py (hunk 4 above)
Comment (casperdcl) [reply tests/func/test_api.py:60]: the only other classes that have that are `tests.remotes.HTTP` -> `TestHTTPRemote` (which doesn't actually look used?) and `tests.remotes.SSHMocked`

Comment (efiop) [reply tests/func/test_api.py:60]: @casperdcl That's some primitive caching that was added. All other get_url's just generate a new one each time, we could try to do the same.

File: tests/func/test_api.py (hunk 5 above)
Comment (casperdcl) [reply tests/func/test_api.py:13]: yes wasn't sure if the other tests would pass - wanted to debug individually but changed now.

**Comment by casperdcl** [comment]
hmm tests failing with

```
2020-05-31 19:42:01,222 ERROR: failed to push data to the cloud - GDrive remote auth failed with credentials in 'GDRIVE_CREDENTIALS_DATA'.
Backup first, remove of fix them, and run DVC again.
It should do auth again and refresh the credentials.
```

https://travis-ci.com/github/iterative/dvc/jobs/341956240

File: dvc/remote/gdrive.py (hunk 2 above)
Comment (shcheklein) [reply dvc/remote/gdrive.py:399]: it should close if just `return` from the block ... we should check the logic carefully here

File: dvc/remote/gdrive.py (hunk 3 above)
Comment (shcheklein) [reply dvc/remote/gdrive.py:402]: @efiop what do you think?

File: tests/remotes.py [hunk 6]
```diff
@@ -141,10 +142,13 @@ class GDrive:
     def should_test():
         return os.getenv(GDriveRemote.GDRIVE_CREDENTIALS_DATA) is not None
 
-    def get_url(self):
-        if not getattr(self, "_remote_url", None):
```
Comment (shcheklein) [inline tests/remotes.py:145]: removing this will break `TestGDriveRemoteCLI`  most likely

Comment (shcheklein) [reply tests/remotes.py:145]: in general I believe we should probably introduce caching of get_url in for other remotes as well - it is strange to return different values within one test

File: tests/func/test_api.py [hunk 7]
```diff
@@ -8,9 +8,9 @@
 from dvc.main import main
 from dvc.path_info import URLInfo
 from dvc.utils.fs import remove
-from tests.remotes import GCP, HDFS, OSS, S3, SSH, Azure, Local
+from tests.remotes import GCP, HDFS, OSS, S3, SSH, Azure, GDrive, Local
 
-remote_params = [S3, GCP, Azure, OSS, SSH, HDFS]
+remote_params = [S3, GCP, Azure, GDrive, OSS, SSH, HDFS]
```
Comment (shcheklein) [inline tests/func/test_api.py:13]: we have a lot of tests for functions besides open here - do we expect all of them to pass? we'll need to implement more stuff then

File: dvc/remote/gdrive.py (hunk 3 above)
Comment (efiop) [reply dvc/remote/gdrive.py:402]: @shcheklein Sorry or the delay. I think we could leave it with the assert for now, same as in other places.

File: tests/remotes.py (hunk 6 above)
Comment (efiop) [reply tests/remotes.py:145]: @shcheklein It actually shouldn't, because other remotes do that too and just reuse the results. Need to check maybe there is something wrong or special about gdrive...

File: tests/func/test_api.py (hunk 7 above)
Comment (efiop) [reply tests/func/test_api.py:13]: @shcheklein All the tests should work, there is nothing new required from GdriveRemotes. There might be some issues with Gdrive itself though (creds or something), that differentiates it from other remotes.

File: tests/remotes.py (hunk 6 above)
Comment (shcheklein) [reply tests/remotes.py:145]: @efiop I think they just call it once, in case of GDrive it just happened that we need it in multiple places. And in general, it's def not a good and safe interface - this value should be cached one way or another and stay the same during the test.

File: tests/func/test_api.py (hunk 7 above)
Comment (shcheklein) [reply tests/func/test_api.py:13]: Even it works out of the box, some things do not make much sense - e.g. what will `dvc.api.url` return for the object? Some URL that no tool (except DVC) can work with?

Re credentials - should be fine (we use an env var to set them, so I would expect external repo to pick it up as well). As I mentioned though, GDrive remote requires some special setup (e.g. it does not create a directory for you).

**Comment by efiop** [comment]
Adjusted GDrive to have a `create_dir` method that is needed before you can use GDrive remote to push to it. Need to adjust those remote test classes to call it for the setup. @casperdcl could you take a look?

File: tests/func/test_api.py (hunk 7 above)
Comment (shcheklein) [reply tests/func/test_api.py:13]: regarding `get_url`, I think that's how the value should look like:

Screen Shot 2020-06-04 at 5 00 38 PM

(`id=` is the file ID that we can get via API)

File: dvc/remote/gdrive.py (hunk 2 above)
Comment (casperdcl) [reply dvc/remote/gdrive.py:399]: upstream issue: https://github.com/iterative/PyDrive2/issues/44

File: tests/func/test_api.py (hunk 7 above)
Comment (casperdcl) [reply tests/func/test_api.py:13]: @shcheklein is this comment outdated since rebasing on master? Or does `tests/remotes.py` still need updating?

Comment (shcheklein) [reply tests/func/test_api.py:13]: @casperdcl primarily gdrive logic should be updated I think, the one that generates URLs - right now they probably look like `gdrive://root/....` which is not very meaningful - no tools understand `gdrive://` URLs . Also, it probably returns path not ID, which is potentially not unique.

Ideally we would have an URL like on the image above (you can click and download in browser). Or it might be somethings else.

The bottom line here is that `get_url` for GDrive needs probably some research on its own and can be a separate PR, and for now we probably should throw NotImplemented.

Unless you guys see some easy and fast implementation.

Comment (casperdcl) [reply tests/func/test_api.py:13]: I think you're talking about `api.get_url` rather than `tests.remotes.GDrive.get_url`. Pretty sure the latter is *meant* to work as-is after @efiop's changes but the tests are still failing requiring 2FA/confirmation codes.

Comment (efiop) [reply tests/func/test_api.py:13]: @casperdcl You need to `create_dir`, `get_url` is not returning an already created dir.

Comment (casperdcl) [reply tests/func/test_api.py:13]: Hmm. Any reason we can't move `create_dir` to within `get_url`?

Comment (efiop) [reply tests/func/test_api.py:13]: @casperdcl Could do that, sure. Will require some work to make it do that without first passing it a dvc instance.

Comment (casperdcl) [reply tests/func/test_api.py:13]: well the only other way I see is to rework the current pytest fixture (currently it hides the `GDrive` object so there's no way to call `create_dir`)

Comment (efiop) [reply tests/func/test_api.py:13]: @casperdcl Both work. Choose the easiest one, we'll be revisiting this later anyway.

Comment (casperdcl) [reply tests/func/test_api.py:13]: tried a simple change along the same lines as `setup_remote` but still failing with the same credentials error https://travis-ci.com/github/iterative/dvc/jobs/348374056#L7952

Comment (efiop) [reply tests/func/test_api.py:13]: @casperdcl does it work locally?

File: tests/func/test_api.py [hunk 8]
```diff
@@ -58,6 +63,7 @@ def test_get_url_requires_dvc(tmp_dir, scm):
 
 @pytest.mark.parametrize("remote_url", all_remote_params, indirect=True)
 def test_open(remote_url, tmp_dir, dvc):
+    ensure_dir(dvc, remote_url)
     run_dvc("remote", "add", "-d", "upstream", remote_url)
```
Comment (efiop) [inline tests/func/test_api.py:67]: The reason why it doesn't work is because gdrive remote needs additional configuration. See TestGDriveRemoteCLI in test_data_cloud.py

File: tests/func/test_api.py [hunk 9]
```diff
@@ -72,6 +122,7 @@ def test_open(remote_url, tmp_dir, dvc):
 @pytest.mark.parametrize("remote_url", all_remote_params, indirect=True)
 def test_open_external(remote_url, erepo_dir, setup_remote):
     setup_remote(erepo_dir.dvc, url=remote_url)
+    ensure_dir_scm(erepo_dir.dvc, remote_url)
```
Comment (casperdcl) [inline tests/func/test_api.py:125]: @efiop I've added `ensure_dir_scm` here based off `setup_remote`.

<details>

- Using `ensure_dir` instead of `ensure_dir_scm`, there's
`ERROR: configuration error - config file error: remote 'upstream' doesn't exists.`
- Using `run_dvc("remote", "add", ...)` instead results in
`ERROR: configuration error - config file error: remote 'upstream' already exists. Use `-f|--force` to overwrite it.`

</details>

File: dvc/exceptions.py [hunk 10]
```diff
@@ -222,9 +222,12 @@ def __init__(self, etag, cached_etag):
 
 
 class FileMissingError(DvcException):
-    def __init__(self, path):
+    def __init__(self, path, hint=None):
         self.path = path
-        super().__init__(f"Can't find '{path}' neither locally nor on remote")
+        self.hint = "" if hint is None else f". {hint}"
```
Comment (efiop) [inline dvc/exceptions.py:227]: self.hint is not used

Comment (casperdcl) [reply dvc/exceptions.py]: oops really silly. one of *those* days.

**Comment by efiop** [comment]
Thanks @casperdcl ! For the record: s3 test is failing for unrelated reasons.

**Comment by efiop** [comment]
@casperdcl Please don't forget about the docs :slightly_smiling_face:
